github: add pull request and issue template files

aabadie opened the pull request:
This is related to the discussion about how to improve initial quality state of contributed PRs, especially from newcomers.

GitHub proposes a feature for setting up an initial PR comment, using a template file. See their documentation [here](https://github.com/blog/2111-issue-and-pull-request-templates) and [here](https://help.github.com/articles/creating-a-pull-request-template-for-your-repository/)

Any comment about the content of this template is welcomed.

14mRh4X0r commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,24 @@
+
+### Preliminary checks
+
+The RIOT community takes a lot of care to contributions quality.
```
It might be better to rephrase this as "The RIOT community cares a lot about code quality."

14mRh4X0r commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,24 @@
+
+The RIOT community takes a lot of care to contributions quality.
+Thus before describing what your contribution is about, we would like you to
+verify the following checklist:
```
Suggestions:
- 'Thus' → 'Therefore,'
- 'verify the following checklist' → 'check the following'

14mRh4X0r commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,24 @@
+The RIOT community takes a lot of care to contributions quality.
+Thus before describing what your contribution is about, we would like you to
+verify the following checklist:
+- [ ] Have you read once the
```
I would either remove "once" or put it at the end of the sentence (" at least once?")

aabadie commented:
@14mRh4X0r, comments addressed. Thanks!

14mRh4X0r commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,24 @@
+Thanks for contributing to RIOT !
```
If I remember correctly, there shouldn't be a space between the last word and interpunction at the end of a sentence, so: "Thanks  for contributing to RIOT!" (same for other lines, e.g. with question marks).

14mRh4X0r commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,24 @@
+
+### Contribution description
+
+_Put here the description of your contibution_
```
I'd move 'here' to the end of the sentence.

14mRh4X0r commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,24 @@
+
+### Issue references
+
+_Put here the list of related issues or PRs (if any)_
```
Same here.

14mRh4X0r replied:
Well, between 'PRs' and '(if any)' to be exact.

14mRh4X0r commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,24 @@
+
+### Preliminary checks
+
+The RIOT community cares a lot about code quality..
```
There's a double period here.

aabadie commented:
@14mRh4X0r, addressed your last comments.

kaspar030 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,24 @@
+
+The RIOT community cares a lot about code quality.
+Therefore before describing what your contribution is about, we would like you
+to check the following checklist:
```
maybe a newline here?

kaspar030 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,24 @@
+- [ ] Have you checked that your modifications are compliant with RIOT coding
+conventions?
+- [ ] Have you rebased your branch against latest RIOT `master` branch (
+`git pull --rebase upstream master`)
```
it is "rebase on sth", not "rebase against" something. also, probably not everyone has "upstream" as remote, so the example command line might not work in that case.

aabadie replied:
indeed, but I didn't know how to simply describe the procedure. I put the minimal list of commands in 'code' block, do you think it's better like this ?
Otherwise we could also add a 'check the github git usage documentation' ?

miri64 commented on .github/PULL_REQUEST_TEMPLATE.md:
````diff
@@ -0,0 +1,29 @@
+```
+$ git remote add upstream git@github.com:RIOT_OS/RIOT.git
+$ git checkout <your branch>
+$ git pull --rebase upstream master
````
For the sake of CLI-newbies: maybe drop the leading `$ `

miri64 commented:
We probably also should update https://github.com/RIOT-OS/RIOT/blob/master/CONTRIBUTING.md and reference it (is this still shown as a link in PRs btw?)

miri64 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,29 @@
+Thanks for contributing to RIOT!
```
Maybe some note that this message and the "Preliminary checks" section are supposed to be removed after reading?

aabadie replied:
I was more thinking at adding a message after the checks saying: "if one of the boxes is not checked, we will ignore your contribution" or something more politically correct but with the same meaning

smlng replied:
I would remove this line, IMHO the template should be short and to the point. Primarily providing structure to help reviewers and testers.

miri64 replied:
Please address @smlng comment

aabadie commented:
> We probably also should update https://github.com/RIOT-OS/RIOT/blob/master/CONTRIBUTING.md and reference it (is this still shown as a link in PRs btw?)

yes, there's a box above the PR initial comment window with a link to the document

OlegHahm commented:
I wonder if there is an easy way to make newcomers aware of a potential impact on memory and/or energy consumption.

smlng commented:
@OlegHahm AFAIR @kaspar030 is working on integrating binary/memory size comparison into Murdock, so CI results will show diffs of PR to (latest) master branch. Energy consumption is difficult, to say the least 😉

smlng commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,42 @@
+
+### Preliminary checks
+
+The RIOT community cares a lot about code quality.
```
this whole section is way to *chatty* for my taste. Why not have checkboxes with short descriptions, and thats it.

Like:

### Preliminary checks

- [ ] code complies to [RIOT coding conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions)
- [ ] no errors reported by `make static-test` (run from repository root)
- [ ] PR is **not** *work-in-progress* (WIP) nor relies on any other PR

Further information can be found in the RIOT [contributing guide](../CONTRIBUTING.md).

aabadie replied:
A welcome message is important for newcomers and dheckboxes are useless when you write the PR description. That's why I removed them.

I also think that it's important to write things correctly (or 'chatty' if you prefer) for people who are not used to open-source contributions as you are.

aabadie commented:
Some update here:
* added an issue template file
* updated the PR template file: the idea is to use html comments to not bloat the initial PR/Issue comment

kYc0o commented:
I think this is getting forgotten, please take a look at your reviews @kaspar030 @smlng.

Personally I think this PR would improve the situation of making good PRs.

miri64 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,54 @@
+to answer the following questions:
+-->
+
+* Have you read the [contributing guide](../CONTRIBUTING.md) (and related wiki
```
I'm not sure this link will resolve as you expect. Also: this link already appears on top of the text box for a new PR in a nice warning yellow. People who ignore that will most likely also ignore this line, so I think it is redundant.

smlng commented:
@aabadie I (very much) support your effort here, but I still encourage to simplify the *preliminary checks* section using checkboxes as I suggested above.

aabadie commented:
> I still encourage to simplify the preliminary checks section using checkboxes as I suggested above.

Ok, I simply removed them in 806236346 to keep something simple. I'm also using html comments: their are displayed while editing the initial comment but hidden in the preview published version.

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,28 @@
+<!--
+If your issue is a usage question, submit it to the user mailing-list:
+users@riot-os.org or devel@riot-os.org
```
"[...] or the developer mailing-list devel@riot-os.org."

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,28 @@
+-->
+
+#### Description
+<!-- Example: Cannot build gnrc_networking application for samr21-xpro board -->
```
Period at end of sentence missing.

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,28 @@
+<!--
+Try to describe as precisely as possible here the steps required to reproduce
+the issue. Here you can also describe your hardware configuration, the network
+setup, etc
```
Period at end of sentence missing.

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,28 @@
+#### Versions
+<!--
+Operating system: Mac OSX, Linux, Vagrant VM
+Build environment: GCC, CLang versions
```
Maybe add a hint to `./dist/tools/ci/print_toolchain_versions.sh`?

miri64 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,25 @@
+The RIOT community cares a lot about code quality.
+Therefore, before describing what your contribution is about, we would like
+you to make sure that your modifications are compliant with the RIOT
+[coding conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions).
```
This isn't parsed as markdown, so using markdown syntax here might be more confusing than helpful.

miri64 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,25 @@
+### Contribution description
+
+<!--
+_Put here the description of your contribution:_
```
Dito

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,28 @@
+-->
+
+#### Expected results
+<!-- Example: The gnrc_networking application builds on samr21-xpro.-->
```
minor nit-pick: there is no space between `.` and `-->`, but below there is.

kaspar030 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,25 @@
+Example: Fixes #1234. See also #5678.
+Please use keywords (e.g., Fixes) to create link to the issues or pull requests
+you resolved, so that they will automatically be closed when your pull request
+is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
```
final dot missing

aabadie commented:
@miri64 @kaspar030, I addressed your comments. Maybe some wording could also be improved.

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,30 @@
+<!--
+If your issue is a usage question, submit it to the user mailing-list
```
maybe add a "please" before submit ;-)?

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,30 @@
+-->
+
+#### Expected results
+<!-- Example: The gnrc_networking application builds on samr21-xpro. -->
```
I stumble now the second time over this sentence and for the second time I only realized after the second read-through, that this is an example. Don't know why though with this particular line. The other examples are not that much a problem.

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,30 @@
+
+#### Description
+<!-- Example: Cannot build gnrc_networking application for samr21-xpro
+board. -->
```
Other multli-line comments in your files have the following format.

```md
<!--
Example: Cannot build gnrc_networking application for samr21-xpro board.
-->
```

(I know this doesn't make it a multiline comment anymore, but I think for consistency this is better.

miri64 replied:
(other alternative would be to maybe make it a oneline comment altogether.

miri64 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,25 @@
+Example: Fixes #1234. See also #5678.
+Please use keywords (e.g., Fixes) to create link to the issues or pull requests
+you resolved, so that they will automatically be closed when your pull request
+is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
```
PRs usually don't get closed when a referring PR gets merged (i.e. the first sentence just applies to issues). Maybe adapt the wording appropriately here to not be misleading.

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,30 @@
+<!--
+If your issue is a usage question, submit it to the user mailing-list
+users@riot-os.org or the developer mailing-list devel@riot-os.org.
```
Maybe also mention `security@riot-os.org` for reporting vulnerabilities.

miri64 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -15,10 +18,13 @@ setup, etc.
 #### Expected results
-<!-- Example: The gnrc_networking application builds on samr21-xpro. -->
+<!--
+Example: The gnrc_networking application builds on samr21-xpro.-->
```
Line-break missing before `-->`

miri64 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -18,8 +18,8 @@ Put here the description of your contribution:
-you resolved, so that they will automatically be closed when your pull request
-is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
+Examples: Fixes #1234. See also #5678. Depends on PR #9876.
+Please use keywords (e.g., fixes, resolve) with link the issues you resolved,
```
"with the link to the issues […]"? In any way: This sentence is broken now.

miri64 commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,25 @@
+<!--
+Examples: Fixes #1234. See also #5678. Depends on PR #9876.
+Please use keywords (e.g., fixes, resolve) with link the issues you resolved,
+this way they will be automatically closed when your pull request is merged.
```
This sentence still doesn't make sense...

smlng replied:
not very helpful comment, though 🤔 But @miri64 has a point, the following mind read better:

* Please use keywords (e.g., fixes, resolve) ~~with~~[and] link the issue you resolved [...], *or*
* Please use keywords (e.g., fixes, resolve) with links [to] the issues you resolved [...]

I would also put the Examples below this sentence.

miri64 replied:
@smlng I had similar proposals already in https://github.com/RIOT-OS/RIOT/pull/7650#discussion_r155643015 ;-)

smlng replied:
oops sorry, hadn't re-read the whole discussion ...

aabadie commented:
@miri64 @smlng I addressed the remaining comment and directly squashed

aabadie commented:
@smlng @kaspar030 do you ACK ?

kaspar030 commented:
Let's go!
